### ISSUE TITLE
clickLink should work correctly for internal links when testing a `Browser.application`

### ISSUE THREAD

**Issue by BrianHicks** [avh4/elm-program-test#107]
I am testing a `Browser.application` which interprets `Browser.UrlRequest`'s `Internal` case for navigation instead of using `Events.onClick` handlers. When I call `clickLink "Foo" (Route.toString Route.Foo)` elm-program-test reports that the program has exited instead of navigating.

I can work around this for now by either manually sending a navigation `Msg` in the test or using `Events.onClick` as well as the href, but it'd be nice if elm-program-test would intercept the links in the same way the real runtime does. 😄

**Comment by avh4** [comment]
Implementation note: hopefully this can be done by having `clickLink` delegate to `routeChangeHelper` so that they share the same internal mechanism.

**Comment by nunntom** [comment]
@avh4 Do you mind if I have a go at a PR for this?

**Comment by avh4** [comment]
> Do you mind if I have a go at a PR for this?

@nunntom Sure, got for it!  Though I'd only suggest that if you think you'll have time for some possible back-and-forth on the code review, since I'm a little bit particular about the code in here.  Also, FYI, I'm pretty sure I took at look at this a few months ago and it turned out to be a little bit harder than I first thought so I put it on hold at the time.

Also the recommended workaround at the moment would be to expose your Msg constructors and use `ProgramTest.update` to directly send in the `Msg` that clicking the link would have generated (and possibly doing a `ProgramTest.ensureViewHas [ tag "a", href expectedUrl ]` just before that).

**Comment by nunntom** [comment]
@avh4 Thanks. I'm very happy to spend as much time as necessary on back and forth in code review. I totally understand you being particular about the code.

I've actually already made the PR. Think of it more as a proof of concept at this stage. Basically I noticed that onUrlRequest was not implemented yet in createHelper, so I added it and made a new helper in TestState called urlRequestHelper which calls update with the onUrlRequest msg. The tricky part is Url.Extra.resolve which needs to be properly implemented to handle relative URLs with query and/or fragment. Really this function should probably return a Maybe URL, but since it is used in ChangedPage I felt it was out of scope to make decisions about that at this point. 

From my testing it all seems to work as expected but perhaps you know something I don't about why this approach won't work. I've added some tests to the test suite to demonstrate. I also mentioned in the description another issue not yet reported where using pushUrl with a relative URL with query and/or fragment produces a URL that doesn't parse correctly with Url.Parser. There's a test to demonstrate this and it would be fixed again by properly implementing Url.Extra.resolve.

PR here: https://github.com/avh4/elm-program-test/pull/166
